File: tests/e2e/helpers/check-invariants.ts
import { execSync } from 'child_process';
import * as path from 'path';

/** Path to the check-invariants binary (release build). */
const BINARY = path.resolve(__dirname, '../../../target/release/check-invariants');

/**
 * Run the CURD conservation invariant check.
 *
 * Connects to the Freenet node, GETs all user contracts, and verifies:
 * 1. Each contract's cached balance matches its ledger replay
 * 2. Total CURD across all contracts == 1,000,000 (SYSTEM_FLOAT)
 *
 * Throws if the check fails.
 */
export function checkInvariants(label: string): void {
  try {
    const output = execSync(`"${BINARY}" "${label}" --soft`, {
      timeout: 120_000,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    if (output.trim()) {
      console.log(output.trim());
    }
  } catch (err: any) {
    const stderr = err.stderr?.toString() || '';
    const stdout = err.stdout?.toString() || '';
    throw new Error(
      `CURD invariant check failed at "${label}":\n${stdout}\n${stderr}`
    );
  }
}

File: tests/node-integration/src/bin/check_invariants.rs
//! Standalone CURD conservation invariant checker.
//!
//! Connects to a Freenet node, dynamically discovers all user contracts from
//! the system root's ledger, and verifies:
//! 1. Each contract's `balance_curds` == `derive_balance()` (cache matches ledger)
//! 2. Sum of all balances == SYSTEM_FLOAT (1,000,000 CURD)
//!
//! Exits 0 on success, 1 on failure. Output goes to stdout.
//!
//! Usage:
//!   check-invariants [LABEL] [--port PORT]
//!
//! LABEL is an optional human-readable context string (e.g. "e2e test 09 pre").
//! --port: Freenet node WebSocket port (default: 3001, the gateway).

use cream_node_integration::harness::{check_curd_conservation, user_contract_key_for};
use cream_node_integration::{connect_to_node_at, node_url, wait_for_get};
use cream_common::user_contract::UserContractState;
use std::time::Duration;

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();

    let port = args
        .windows(2)
        .find(|w| w[0] == "--port")
        .and_then(|w| w[1].parse::<u16>().ok())
        .unwrap_or(3001);

    let dump = args.iter().any(|a| a == "--dump");

    if dump {
        // Dump mode: print ledger entries for a named contract
        let name = args.get(1).filter(|a| *a != "--dump" && *a != "--port")
            .cloned().unwrap_or("system_root".to_string());
        let key = user_contract_key_for(&name);
        let url = node_url(port);
        let mut api = connect_to_node_at(&url).await;
        let bytes = wait_for_get(&mut api, *key.id(), Duration::from_secs(30)).await
            .expect("GET contract");
        let state: UserContractState = serde_json::from_slice(&bytes).unwrap();
        println!("Contract: {} (port {})", name, port);
        println!("balance_curds: {}", state.balance_curds);
        println!("derive_balance(): {}", state.derive_balance());
        println!("checkpoint_balance: {}", state.checkpoint_balance);
        println!("Ledger ({} entries):", state.ledger.len());
        for tx in &state.ledger {
            println!("  {:?} {:>6} {:30} sender={:20} receiver={:20} tx_ref={}",
                tx.kind, tx.amount, tx.description, tx.sender, tx.receiver, tx.tx_ref);
        }
        return;
    }

    let label = args
        .get(1)
        .filter(|a| *a != "--port")
        .cloned()
        .unwrap_or_else(|| "standalone".to_string());

    check_curd_conservation(&label, port).await;
}
